p2p/http/libp2phttp.go

```diff
@@ -0,0 +1,688 @@
+	for _, opt := range opts {
+		err := opt(h)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return h, nil
+}
+
+func (h *HTTPHost) Addrs() []ma.Multiaddr {
+	<-h.httpTransport.waitingForListeners
+	return h.httpTransport.listenAddrs
+}
+
+var ErrNoListeners = errors.New("nothing to listen on")
+
+// Serve starts the HTTP transport listeners. Always returns a non-nil error.
+// If there are no listeners, returns ErrNoListeners.
+func (h *HTTPHost) Serve() error {
+	closedWaitingForListeners := false
+	defer func() {
+		if !closedWaitingForListeners {
+			close(h.httpTransport.waitingForListeners)
+		}
+	}()
+
+	if len(h.httpTransport.requestedListenAddrs) == 0 && h.streamHost == nil {
+		return ErrNoListeners
+	}
+
+	h.httpTransport.listeners = make([]net.Listener, 0, len(h.httpTransport.requestedListenAddrs)+1) // +1 for stream host
+
+	streamHostAddrsCount := 0
+	if h.streamHost != nil {
+		streamHostAddrsCount = len(h.streamHost.Addrs())
+	}
+	h.httpTransport.listenAddrs = make([]ma.Multiaddr, 0, len(h.httpTransport.requestedListenAddrs)+streamHostAddrsCount)
+
+	errCh := make(chan error)
+
+	if h.streamHost != nil {
+		listener, err := StreamHostListen(h.streamHost)
+		if err != nil {
+			return err
+		}
+		h.httpTransport.listeners = append(h.httpTransport.listeners, listener)
+		h.httpTransport.listenAddrs = append(h.httpTransport.listenAddrs, h.streamHost.Addrs()...)
+
+		go func() {
+			errCh <- http.Serve(listener, &h.rootHandler)
+		}()
+	}
+
+	closeAllListeners := func() {
+		for _, l := range h.httpTransport.listeners {
+			l.Close()
+		}
+	}
+
+	for _, addr := range h.httpTransport.requestedListenAddrs {
+		parsedAddr := parseMultiaddr(addr)
+		// resolve the host
+		ipaddr, err := net.ResolveIPAddr("ip", parsedAddr.host)
+		if err != nil {
+			closeAllListeners()
+			return err
+		}
+
+		host := ipaddr.String()
+		l, err := net.Listen("tcp", host+":"+parsedAddr.port)
+		fmt.Println("HTTPHost.Serve", err)
+		if err != nil {
+			closeAllListeners()
+			return err
+		}
+		h.httpTransport.listeners = append(h.httpTransport.listeners, l)
+
+		// get resolved port
+		_, port, err := net.SplitHostPort(l.Addr().String())
+		if err != nil {
+			closeAllListeners()
+			return err
+		}
+
+		var listenAddr ma.Multiaddr
+		if parsedAddr.useHTTPS && parsedAddr.sni != "" && parsedAddr.sni != host {
+			listenAddr = ma.StringCast(fmt.Sprintf("/ip4/%s/tcp/%s/tls/sni/%s/http", host, port, parsedAddr.sni))
+		} else {
+			scheme := "http"
+			if parsedAddr.useHTTPS {
+				scheme = "https"
+			}
+			listenAddr = ma.StringCast(fmt.Sprintf("/ip4/%s/tcp/%s/%s", host, port, scheme))
+
+		}
+
+		h.httpTransport.listenAddrs = append(h.httpTransport.listenAddrs, listenAddr)
+
+		if parsedAddr.useHTTPS {
+			go func() {
+				srv := http.Server{
+					Handler:   &h.rootHandler,
+					TLSConfig: h.httpTransport.tlsConfig,
+				}
+				errCh <- srv.ServeTLS(l, "", "")
+			}()
+		} else {
+			go func() {
+				errCh <- http.Serve(l, &h.rootHandler)
+			}()
```

Comment:
I'm thinking about how we'd support a proxy (like nginx) that terminating SSL and communicating over HTTP.

Maybe it's better to make that a bit harder to avoid users broadcasting unencrypted HTTP by accident.